src/MessageHandler/SearchNoHitsMessageHandler.php

```diff
@@ -45,151 +45,312 @@ public function __construct(
+
+            if (IdentifierType::PID === $message->getIdentifierType()) {
+                // For PID's of type '-katalog' we might not have access to them
+                // in the datawell through our search profile. We try to guess the
+                // faust from the PID and search by that to instead.
+                $found = $this->mapCatalogIdentifier($identifier);
+            }
+
+            if (!$found) {
+                // Try to search the data well and match source entity. If the 'source'
+                // have been added by the vendor before it's available in the datawell
+                // it will not have been indexed.
+                $material = $this->searchService->search($identifier, $message->getIdentifierType());
+                $found = $this->mapDatawellSearch($material);
+            }
+
+            if (!$found && isset($material)) {
+                // Some vendors can return a potential URL for the cover given an identifier.
+                // Get a list of these from the matched identifiers in the datawell.
+                $unverifiedVendorImageItems = $this->getUnverifiedVendorImageItems($material);
+
+                $found = $this->processUnverifiedImageItems($unverifiedVendorImageItems);
+            }
+
+            if (!$found) {
+                $this->metricsService->counter('no_hit_failed', 'No-hit mapping not found', 1, ['type' => 'nohit']);
+            }
+        } catch (\Exception $exception) {
+            $this->metricsService->counter('no_hit_katelog_error', 'No-hit katelog error', 1, ['type' => 'nohit']);
+
+            $this->logger->error('Exception: '.$exception::class, [
+                'service' => self::class,
+                'message' => $exception->getMessage(),
+                'identifier' => $message->getIdentifier(),
+            ]);
+        }
+    }
+
+    /**
+     * Get unverified image items from vendors for a given material.
+     *
+     * @param Material $material
+     *
+     * @return UnverifiedVendorImageItem[]
      *
-     * @throws MaterialTypeException
-     * @throws OpenPlatformAuthException
-     * @throws MaterialTypeException
-     * @throws OpenPlatformAuthException
-     * @throws OpenPlatformSearchException
-     * @throws Exception
+     * @throws UnknownVendorServiceException
+     * @throws UnsupportedIdentifierTypeException
      */
-    public function __invoke(SearchNoHitsMessage $message)
+    private function getUnverifiedVendorImageItems(Material $material): array
     {
-        $identifier = $message->getIdentifier();
+        $items = [];
+
+        foreach ($material->getIdentifiers() as $identifier) {
+            /** @var SupportsSingleIdentifierInterface $vendor */
+            foreach ($this->singleIdentifierVendors as $vendor) {
+                if ($vendor->supportsIdentifierType($identifier->getType())) {
+                    $items[] = $vendor->getUnverifiedVendorImageItem($identifier->getId(), $identifier->getType());
```

Comment:
Is these `singleIdentifierVendors` order by weight in the iterator? If so could processing be stop after first found item?

Reply:
The implementation follow the same logic as for importers. We check all vendors and save the source even if a different vendor have a source for the same identifier. Then leave it to the `IndexMessageHandler` to handle rank when updating the index. This is by design to keep the logic that same as for the importers flow.

As for havng them ordered: They are injected by the DI container in no particular order. Maybe we can set a `rank`of `priority` in `services.yaml`but this would duplicate the database rank. Alternatively we would need to sort them before looping over them by calling `$vendor->getRank()` if we decide to change the above logic for this use case.